python/pyspark/ml/model_cache.py

```diff
@@ -0,0 +1,44 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from collections import OrderedDict
+from threading import Lock
+from typing import Callable, Optional
+from uuid import UUID
+
+
+class ModelCache:
+    """Cache for model prediction functions on executors."""
+
+    _models: OrderedDict[UUID, Callable] = OrderedDict()
+    _capacity: int = 8
```

Comment:
I'm a bit concerned about the cache usage. The default behavior shouldn't cause any memory leak, which is not the case here. Couple options:

* Recommend using broadcast if users want to fast load/re-load models.
* Make the caching behavior configurable (spark conf) and leave the default to no caching. And document the potential memory leak issues clearly.

Reply:
From offline discussion:

* Set the default capacity to 3. Later we might make it configurable via Spark conf if needed.
* Document the behavior / trade-offs clearly. Link to https://github.com/apache/spark/blob/77339dc6a49d1d9d2a7a3aae966610acbe1a5d6e/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala#L395

Reply:
Changed default capacity to 3 and updated docs.